Add dev fn to render dashboards with static-viz renderers

adam-james-v opened the pull request:
<img width="1465" alt="image" src="https://github.com/metabase/metabase/assets/21064735/91a1b432-defa-4281-8e60-997ad878245a">

Preview a Dashboard's exports by using the endpoint `/api/pulse/preview_dashboard/:id`

Run `(dev.render-png/render-dashboard-to-html-and-open 1)` to render the dashboard with id 1 to a handy html file which will open in your default browser.

This file contains renders for each dashcard in the dashboard 3 ways:
 - as a png, like you would see in Slack or in an email body (if it's a chart)
 - as html/svg, like you'd see in the email body (for tables).. might also be helpful to inspect the svg output from the graaljs interpreter
 - 10 row table representing what the .csv attachment for the dashcard would look like


Todo:
 - [x] avoid with-redefs to pass the linter
 - [x] add a preview endpoint that can be used so that you don't need to run a Clojure repl to get the same result.
 - [x] demo screenshots and/or recording
 - [ ] make CSV preview work with the endpoint too (not just in dev ns)

adam-james-v committed:
Add dev fn to render dashboards with static-viz renderers

Run `(dev.render-png/render-dashboard-to-html 1)` to render the dashboard with id 1 to a handy html file.

This file will render each dashcard in the dashboard 3 ways:
 - as a png, like you would see in Slack or in an email body (if it's a chart)
 - as html/svg, like you'd see in the email body (for tables).. might also be helpful to inspect the svg output from
 the graaljs interpreter
 - 10 row table representing what the .csv attachment for the dashcard would look like

Note: to get this branch working properly, there are a couple lines commented out to pass the linter... it's a work in
progress and will be cleaned up for better use.

Todo:
 - [ ] avoid with-redefs to pass the linter
 - [ ] add a preview endpoint that can be used so that you don't need to run a Clojure repl to get the same result.

adam-james-v committed:
Add dev fn to render dashboards with static-viz renderers

Run `(dev.render-png/render-dashboard-to-html 1)` to render the dashboard with id 1 to a handy html file.

This file will render each dashcard in the dashboard 3 ways:
 - as a png, like you would see in Slack or in an email body (if it's a chart)
 - as html/svg, like you'd see in the email body (for tables).. might also be helpful to inspect the svg output from
 the graaljs interpreter
 - 10 row table representing what the .csv attachment for the dashcard would look like

Note: to get this branch working properly, there are a couple lines commented out to pass the linter... it's a work in
progress and will be cleaned up for better use.

Todo:
 - [ ] avoid with-redefs to pass the linter
 - [ ] add a preview endpoint that can be used so that you don't need to run a Clojure repl to get the same result.

adam-james-v committed:
Eliminate with-redefs to allow csv 'render'

adam-james-v committed:
Eliminate with-redefs to allow csv 'render'

adam-james-v committed:
Make a 'preview' ns to power `api/pulse/preview_dashboard/:id`

adam-james-v committed:
Make a 'preview' ns to power `api/pulse/preview_dashboard/:id`

adam-james-v committed:
Make dev.render-png dashboard preview fns the same as the preview ns

adam-james-v committed:
Make dev.render-png dashboard preview fns the same as the preview ns

adam-james-v committed:
Dashboard Subscription preview ns now also uses csv attachment code

This change now uses the csv attachment code to include an html table representing what we expect to see from csv exports.

adam-james-v committed:
Dashboard Subscription preview ns now also uses csv attachment code

This change now uses the csv attachment code to include an html table representing what we expect to see from csv exports.

adam-james-v committed:
Merge branch 'render-dashboard-with-static-viz-pipeline' of https://github.com/metabase/metabase into render-dashboard-with-static-viz-pipeline

adam-james-v committed:
Merge branch 'master' into render-dashboard-with-static-viz-pipeline

adam-james-v committed:
Merge branch 'master' into render-dashboard-with-static-viz-pipeline

adam-james-v committed:
Merge branch 'render-dashboard-with-static-viz-pipeline' of https://github.com/metabase/metabase into render-dashboard-with-static-viz-pipeline

markbastian commented on dev/src/dev/render_png.clj:
```diff
@@ -85,29 +90,123 @@
-                                       :hidden-columns      {:hide [0 2]}})
-      :viz-tree
-      open-hiccup-as-html))
+(def ^:private execute-dashboard #'pulse/execute-dashboard)
```
I usually just call private functions as regular aliased functions -- `(#'pulse/execute-dashboard ...)` inline. Were you just defining this here for convenience?

markbastian commented on dev/src/dev/render_png.clj:
```diff
@@ -85,29 +90,123 @@
+    (with-open [w (io/writer tmp-file)]
+      (.write w ^String html-str))
+    (.deleteOnExit tmp-file)
+    (open tmp-file)))
```
I love all the new dev functions here.

markbastian commented on src/metabase/api/pulse.clj:
```diff
@@ -275,6 +276,18 @@
                           render/*include-buttons* true]
                   (render/render-pulse-card-for-display (metabase.pulse/defaulted-timezone card) card result))]])}))
 
```
My only big question here is if we want to enable this publicly, because then it's a supported API. I'd much rather expose it only if the dev profile is active. I don't see why a customer would want this, but it's exceptionally valuable for development. WDYT?

markbastian commented on src/metabase/pulse/preview.clj:
```diff
@@ -0,0 +1,163 @@
+   {:border          "1px solid black"
+    :border-collapse "collapse"}))
+
+(def ^:private csv-row-limit 10)
```
Might want to make this a query param and pass it in that way.

markbastian commented on src/metabase/pulse/preview.clj:
```diff
@@ -0,0 +1,163 @@
+      "nux" :unix
+      nil))
+
+;; taken from https://github.com/aysylu/loom/blob/master/src/loom/io.clj
```
This fn and ns also feels very dev-y. Is there a way to move this and the defendpoint into our dev folder and conditionally activate it that way?

adam-james-v replied:
removed from the preview ns

markbastian reviewed (changes requested):
I like it all except I think maybe we want to relocate it so it only shows up in dev. Can we load that defendpoint when you run `(dev/start!)`?

adam-james-v committed:
Collect all element styles into a single style tag to use with the server's nonce

This allows the dashboard_preview endpoint to render properly by gathering all of the element styles into a single
style tag, which can then be given the server's nonce value so that CSP doesn't strip the style out.

This is useful for the preview endpoint so that we get an accurate picture of what our tables will look like in
emails, plus it makes the presentation look just a bit more readable.

adam-james-v committed:
Merge branch 'master' into render-dashboard-with-static-viz-pipeline

adam-james-v committed:
Merge branch 'master' into render-dashboard-with-static-viz-pipeline

adam-james-v committed:
dev render_png fns match output used on the endpoint now too.

adam-james-v committed:
Merge branch 'render-dashboard-with-static-viz-pipeline' of https://github.com/metabase/metabase into render-dashboard-with-static-viz-pipeline

adam-james-v committed:
Merge branch 'master' into render-dashboard-with-static-viz-pipeline

adam-james-v committed:
Hickory dependency out of dev into regular deps
